src/common/function/src/scalars/timestamp/to_timezone.rs

```diff
@@ -0,0 +1,422 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use common_query::error::{InvalidFuncArgsSnafu, Result, UnsupportedInputDataTypeSnafu};
+use common_query::prelude::{Signature, Volatility};
+use common_time::timezone::TimeZone;
+use common_time::Timestamp;
+use datatypes::prelude::{ConcreteDataType, Vector};
+use datatypes::types::TimestampType;
+use datatypes::value::Value;
+use datatypes::vectors::{
+    Int64Vector, StringVector, TimestampMicrosecondVector, TimestampMillisecondVector,
+    TimestampNanosecondVector, TimestampSecondVector, VectorRef,
+};
+use snafu::ensure;
+
+use crate::scalars::function::{Function, FunctionContext};
+
+#[derive(Clone, Debug, Default)]
+pub struct ToTimeZoneFunction;
+
+const NAME: &str = "to_timezone";
+
+fn convert_to_timezone(arg: &str) -> Option<TimeZone> {
+    match TimeZone::from_tz_string(arg) {
+        Ok(ts) => ts,
+        Err(_err) => None,
+    }
+}
+
+fn convert_to_timestamp(arg: &Value) -> Option<Timestamp> {
+    match arg {
+        Value::Int64(ts) => Some(Timestamp::from(*ts)),
+        Value::String(ts) => match Timestamp::from_str(ts.as_utf8()) {
+            Ok(ts) => Some(ts),
+            Err(_) => None,
+        },
+        Value::Timestamp(ts) => Some(*ts),
+        _ => None,
+    }
+}
+
+fn process_to_timezone(
+    times: Vec<Option<Timestamp>>,
+    tzs: Vec<Option<TimeZone>>,
+) -> Vec<Option<String>> {
+    times
+        .iter()
+        .zip(tzs.iter())
+        .map(|(time, tz)| match (time, tz) {
+            (Some(time), _) => Some(time.to_timezone_aware_string(tz.clone())),
+            _ => None,
+        })
+        .collect::<Vec<Option<String>>>()
+}
+
+impl Function for ToTimeZoneFunction {
+    fn name(&self) -> &str {
+        NAME
+    }
+
+    fn return_type(&self, _input_types: &[ConcreteDataType]) -> Result<ConcreteDataType> {
```

Comment:
@evenyag @killme2008 if we choose the timestamp type as return type, which one should be returned here:

```rust
pub enum TimestampType {
    Second(TimestampSecondType),
    Millisecond(TimestampMillisecondType),
    Microsecond(TimestampMicrosecondType),
    Nanosecond(TimestampNanosecondType),
}
```

The input type can be string or integer other than timestamp so we cannot infer the type the same as the input timestamp type.

Reply:
Maybe return an error if input types are not timestamps? We should return a timestamp with the same time unit as the input one.

Reply:
This makes sense. But how if users want to `to_timezone(ts_str, tzname)`? They should cast the string to timestamp themselves? And how.